## 互评作业三：基于支持向量机的手写数字识别
本脚本采用支持向量机（SVM）方法对手写数字进行识别。通过对 MNIST 数据集进行处理，划分训练集和测试集。然后构建支持向量机模型，进行模型训练和评估。

姓名：王鸿远 学号：3120220967

代码仓库：https://github.com/HW-VMCL/DataMining

1. 读取 MNIST 数据：加载并处理手写数字图像数据及其对应的标签。

In [4]:
import sklearn
import numpy as np
import struct
## 读取MNIST数据，存为numpy数组，并打印其形状，仅采用官方提供的60000张训练数据。
with open('train-labels.idx1-ubyte','rb') as file:
    magic, n = struct.unpack('>II',file.read(8))
    labels = np.fromfile(file,dtype=np.uint8)
with open('train-images.idx3-ubyte', 'rb') as file:
    magic, num, rows, cols = struct.unpack('>IIII',file.read(16))
    images = np.fromfile(file,dtype=np.uint8).reshape(len(labels), 784)

print(labels.shape)
print(images.shape)

(60000,)
(60000, 784)


2. 划分训练集和测试集：将图像数据按照8:2比例划分为训练集和测试集。

In [5]:
### 利用sklearn中提供的随机切分训练集和测试集的库函数，对数据进行切分，并打印其形状。
from sklearn.model_selection  import train_test_split
x_train,x_test,y_train,y_test = train_test_split(images,labels,test_size=0.2,random_state=816)
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(48000, 784)
(12000, 784)
(48000,)
(12000,)


3. 特征缩放：对图像数据进行归一化处理，缩放到[0,1]范围。

In [8]:
### 打印训练图像和标签的最大值，根据最大值将数据归一化处理。
print(x_train.max(),x_test.max())
print(y_train.max(),y_test.max())
x_train/=255.0
x_test/=255.0

255 255
9 9


4. 构建支持向量机模型：用径向基核训练模型。

In [9]:
from sklearn import svm
model = svm.SVC()
model.fit(x_train,y_train)

SVC()

5. 模型评估：在测试集上进行预测，计算模型的准确率、精确率、召回率、F1值，评估模型性能。

In [10]:
### 模型推理
y_out=model.predict(x_test)
print(y_out.shape)

(12000,)


In [18]:
### 计算准确率、精确率、召回率、F1指标
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score
print('准确率：{:.4f}'.format(accuracy_score(y_test,y_out)))
print('精确率：{:.4f}'.format(precision_score(y_test,y_out, average='macro')))
print('召回率：{:.4f}'.format(recall_score(y_test,y_out, average='macro')))
print('F1：{:.4f}'.format(recall_score(y_test,y_out, average='macro')))

准确率：0.9768
精确率：0.9767
召回率：0.9766
F1：0.9766


6. 结果分析

从实验结果来看，以径向基（rbf）为核函数的SVM在MNIST手写数字识别任务上能取得接近98%的准确率，准确率较高，说明实验设置合理，模型性能良好。